network/ssh/reachable.go

```diff
@@ -0,0 +1,216 @@
+	if hostname != remote.String() {
+		debugName = fmt.Sprintf("%s at %s", hostname, remote.String())
+	}
+	logger.Tracef("checking host key for %s, with key %q", debugName, authKeyForm)
+
+	lookupKey := string(key.Marshal())
+	if len(h.AcceptedKeys) == 0 || h.AcceptedKeys.Contains(lookupKey) {
+		logger.Debugf("accepted host key for: %s", debugName)
+		// This key was valid, so return it, but if someone else was found
+		// first, still exit.
+		select {
+		case h.Accepted <- h.HostPort:
+		case <-h.Stop:
+		}
+		return hostKeyAccepted
+	}
+	logger.Debugf("host key for %s not in our accepted set: use --debug --log-level=TRACE to see actual key", debugName)
+	return hostKeyNotInList
+}
+
+// publicKeysToSet converts all the public key values (eg id_rsa.pub) into
+// their short hash form. Problems with a key are logged at Warning level, but
+// otherwise ignored.
+func publicKeysToSet(publicKeys []string) set.Strings {
+	acceptedKeys := set.NewStrings()
+	for _, pubKey := range publicKeys {
+		// key, comment, options, rest, err
+		sshKey, _, _, _, err := ssh.ParseAuthorizedKey([]byte(pubKey))
+		if err != nil {
+			logger.Warningf("unable to handle public key: %q\n", pubKey)
+			continue
+		}
+		acceptedKeys.Add(string(sshKey.Marshal()))
+	}
+	return acceptedKeys
+}
+
+// Check initiates a connection to HostPort and tries to do an SSH key
+// exchange to determine the preferred public key of the remote host.
+// It then checks if that key is in the accepted set of keys.
+func (h *hostKeyChecker) Check() {
+	defer func() {
+		// send a finished message unless we're already stopped and nobody
+		// is listening
+		select {
+		case h.Finished <- struct{}{}:
+		case <-h.Stop:
+		}
+	}()
+	// TODO(jam): 2017-01-24 One limitation of our algorithm, is that we don't
+	// try to limit the negotiation of the keys to our set of possible keys.
+	// For example, say we only know about the RSA key for the remote host, but
+	// it has been updated to use a ECDSA key as well. We
+	sshconfig := &ssh.ClientConfig{
+		HostKeyCallback: h.hostKeyCallback,
+	}
+	addr := h.HostPort.NetAddr()
+	logger.Debugf("dialing %s to check host keys", addr)
+	conn, err := h.Dialer.Dial("tcp", addr)
+	if err != nil {
+		logger.Debugf("dial %s failed with: %v", addr, err)
+		return
+	}
+	// No need to do the key exchange if we're already stopping
+	select {
+	case <-h.Stop:
+		conn.Close()
+		return
+	default:
+	}
+	logger.Debugf("connected to %s, initiating ssh handshake", addr)
+	// NewClientConn will close the underlying net.Conn if it gets an error
+	client, _, _, err := ssh.NewClientConn(conn, addr, sshconfig)
+	if err == nil {
+		// We don't expect this case, because we don't support Auth,
+		// but make sure to close it anyway.
+		client.Close()
+	} else {
+		// no need to log these two messages, that's already been done
+		// in hostKeyCallback
+		if !strings.Contains(err.Error(), hostKeyAccepted.Error()) &&
+			!strings.Contains(err.Error(), hostKeyNotInList.Error()) {
```

Comment:
So this is just to suppress yet-another log message. I figure it is worth logging a message if SSH fails for reasons I don't understand.
I can't use errors.Cause because the error is being returned to Gocrypto, which then reformats it into its own Error structure using:
fmt.Errorf("ssh: handshake failed: %v", err)
https://github.com/golang/crypto/blob/master/ssh/client.go#L77

I'd love to use Cause, but that's only for Juju errors.